### PR TITLE
Use user primary group if not root

### PR DESCRIPTION
**Pull request by Mart124**
Hi,

This PR modifies startup scripts so that when Bitwarden is started from a user which is not `root`, the group used to launch the services is the user's group, rather than the "default" `docker` group.
This is what we expect on system side.

`docker` group remains the chosen group if Bitwarden is launched from `root` account.

Thank you !

### CODE REVIEW

**Comment by kspearrin** [comment]
I'm not sure I understand why we wouldn't still use the **docker** group for all users? Isn't that a standard for running docker?

**Comment by Mart124** [comment]
Of course, if Bitwarden is launched using `root` user, for security reasons, you create a service user belonging to the `docker` group, which is a good default choice.

However, if system administrator has created a specific service user used to launch Bitwarden, for sure he expects the running group to be this specific user's group, not another "default" `docker` group : for security reasons, system administrator could have put his service user into a specific group to isolate it from other users, such as from the ones belonging to the `docker` group.

**Comment by Mart124** [comment]
i.e., if system administrators wants Bitwarden to run with the `docker` group, he will configure his service user to belong to the `docker` group :)

**Comment by kspearrin** [comment]
Wouldn't it be better to check if the user belongs to the standard docker group first? If so, use **docker**, if not use their primary? user group.

**Comment by Mart124** [comment]
No, a system administrator won't expect the service to run with another group but the service user primary group.
If he wants the service to run with another group, then he expects to be able to configure it somewhere. Or he changes the service user's primary group.

File: scripts/install.sh [hunk 1]
```diff
@@ -29,7 +29,12 @@ then
 fi
 
 LUID="LOCAL_UID=`id -u $USER`"
-LGID="LOCAL_GID=`getent group docker | cut -d: -f3`"
+if [ "$OS" == "lin" -a `id -u $USER` -eq 0 ]
+then
+    LGID="LOCAL_GID=`getent group docker | cut -d: -f3`"
```
Comment (kspearrin) [inline scripts/install.sh]: Should we get rid of this condition now based on conversations from yesterday?

Comment (Mart124) [reply scripts/install.sh]: Yes I'm simplifying all this part.
Unfortunately I was stuck on other things right after our conversations, but now I'm on it 👍

**Comment by Mart124** [comment]
So, this new commit simplifies a lot UID/GID management.

It saves installation UID/GID so that they will be reused when Bitwarden is started.
User/admin could also change them if needed, for example to a specific service user.

If UID/GID are set to 0, then Bitwarden switches to the well-known `nobody/nogroup` user.

No more differences between Linux & Mac, same behavior.

Last command of `entrypoint.sh` files is launched using `exec`,  thus getting ride of all `./entrypoint.sh` lines in `ps` command output.

File: scripts/run.sh [hunk 2]
```diff
@@ -79,58 +71,30 @@ function updateLetsEncrypt() {
...
+        --env-file $ENV_DIR/uid.env bitwarden/setup:$COREVERSION \
+        dotnet Setup.dll -printenv 1 -os $OS -corev $COREVERSION -webv $WEBVERSION
 }
 
 function restart() {
     dockerComposeDown
     dockerComposePull
     updateLetsEncrypt
-    
-    if [ $OS == "lin" ]
```
Comment (kspearrin) [inline scripts/run.sh]: Why was this removed? User can change between restarts. ex. going from root to non-root user.

File: src/Admin/entrypoint.sh [hunk 3]
```diff
@@ -5,52 +5,27 @@
...
-    useradd -r -u $LUID -g $GROUPNAME $USERNAME
-fi
-
-# Make home directory for user
-
-if [ ! -d "/home/$USERNAME" ]
-then
-    mkhomedir_helper $USERNAME
-fi
+groupadd -o -g $LGID $GROUPNAME >/dev/null 2>&1 ||
```
Comment (kspearrin) [inline src/Admin/entrypoint.sh]: what does `-o` do?

File: src/Admin/entrypoint.sh [hunk 4]
```diff
@@ -5,52 +5,27 @@
...
-LUID=${LOCAL_UID:-999}
+# Step down from host root to well-known nobody/nogroup user
 
-# Step down from host root
-
-if [ $LGID == 0 ]
+if [ $LUID -eq 0 ]
 then
-    LGID=999
+    LUID=65534
```
Comment (kspearrin) [inline src/Admin/entrypoint.sh]: So containers will never run as root anymore? Even if run from sudo?

File: scripts/run.sh (hunk 2 above)
Comment (Mart124) [reply scripts/run.sh]: `$LOCAL_UID` and `$LOCAL_GID` are set and saved at installation time in `uid.env`, to the UID/GID of the user performing the installation.
At run time, `uid.env` is loaded, and the UID/GID are used by `entrypoint.sh` files to run the services, but `65534/65534` if saved UID/GID are set to 0 (root).

So if `kyle/kyle` performed the installation, subsequent starts will use `kyle/kyle` to run the services.
If `root` performed the installation, `65534/65534` (`nobody/nogroup`) will be used by default, for security reasons (we don't want services to run as `root` user).

User could also want and manually update the `uid.env` file to switch services to another user.
Think about `root` installing the application. He can then modify the `uid.env` file to switch to whatever service user he has properly configured on the system to run Bitwarden with.
Subsequent starts will then use this specific service user, even if Bitwarden is started from root, at boot time etc...
This is how services generally work on UNIX / Linux.

Overwriting `uid.env` each time application starts is useless, and prevents for example `root` to define a specific service user.

File: src/Admin/entrypoint.sh (hunk 4 above)
Comment (Mart124) [reply src/Admin/entrypoint.sh]: We do not want services to run as root, for security reasons.
`sudo` does nothing more than granting `root` rights, thus allowing calling user (which has sort of switched to `root`) to access docker subsystem (and then to start its containers).
Containers then start services using the UID/GID from the `uid.env` file.

On some systems, adding users to the `docker` group is another solution to allow them to use Docker, thus without needing `sudo` anymore : `usermod -aG docker myuser`

File: scripts/run.sh (hunk 2 above)
Comment (kspearrin) [reply scripts/run.sh]: This will now break if the user is upgrading from a version where `uid.env` did not exist. This functionality was only recently added and these scripts must remain backwards compatible.

Comment (Mart124) [reply scripts/run.sh]: You're right, good catch :
`ERROR: Couldn't find env file: /home/bw/bwdata/env/uid.env`
Let me add the required check / few lines 👍

Comment (Mart124) [reply scripts/run.sh]: This does the trick :
https://github.com/bitwarden/core/pull/292/commits/30721b7a8ea32d6cd79119339fa3630a59334214

File: src/Admin/entrypoint.sh [hunk 5]
```diff
@@ -5,52 +5,27 @@
...
 fi
-
-if [ $LUID == 0 ]
+if [ $LGID -eq 0 ]
```
Comment (kspearrin) [inline src/Admin/entrypoint.sh]: Can we switch these back to `== 0` to be consistent with equality checks in other scripts? I prefer that syntax when possible

File: src/Admin/entrypoint.sh [hunk 6]
```diff
@@ -5,52 +5,27 @@
...
-
-if [ ! -d "/home/$USERNAME" ]
-then
-    mkhomedir_helper $USERNAME
-fi
+groupadd -o -g $LGID $GROUPNAME >/dev/null 2>&1 ||
+groupmod -o -g $LGID $GROUPNAME >/dev/null 2>&1
+useradd -o -u $LUID -g $GROUPNAME -s /bin/false $USERNAME >/dev/null 2>&1 ||
+usermod -o -u $LUID -g $GROUPNAME -s /bin/false $USERNAME >/dev/null 2>&1
+mkhomedir_helper $USERNAME
```
[the remark this answers is not in the document]
Comment (Mart124) [reply src/Admin/entrypoint.sh]: I tested, no :)

File: src/Admin/entrypoint.sh [hunk 7]
```diff
@@ -5,52 +5,27 @@
...
-
-# Make home directory for user
-
-if [ ! -d "/home/$USERNAME" ]
-then
-    mkhomedir_helper $USERNAME
-fi
+groupadd -o -g $LGID $GROUPNAME >/dev/null 2>&1 ||
+groupmod -o -g $LGID $GROUPNAME >/dev/null 2>&1
+useradd -o -u $LUID -g $GROUPNAME -s /bin/false $USERNAME >/dev/null 2>&1 ||
```
Comment (kspearrin) [inline src/Admin/entrypoint.sh]: Why is the `-s /bin/false` necessary? Not sure I understand what that is doing here, even after reading the man page on `-s`.

Comment (Mart124) [reply src/Admin/entrypoint.sh]: When you add a user to a system, you give it a login shell, which will be used when this user will connect to the system.
It could be `bash`, `sh`, `ksh`...
Service users are not intended to login, thus we usually give us a non-shell, such as `/bin/false` or `/usr/sbin/nologin`.
This is the last parameter of the `/etc/passwd` file.
I then added `-s /bin/false` to the command, to be consistent, without there was no shell set for the user.

File: src/Admin/entrypoint.sh (hunk 5 above)
Comment (Mart124) [reply src/Admin/entrypoint.sh]: The correct syntax with `[` `]` for integer comparison is `-eq` (see `man test`).
Using `bash`, we should rather use `[[` `]]` which are built-in, compared to `[` `]` which are external commands.
with `[[` `]]` we use `==`.
I however sticked to `[` `]`, as you first used them :)

File: util/Setup/EnvironmentFileBuilder.cs [hunk 8]
```diff
@@ -170,9 +170,6 @@ private void Build()
             }
 
             Helpers.Exec("chmod 600 /bitwarden/env/mssql.override.env");
-
-            // Empty uid env file. Only used on Linux hosts.
-            using(var sw = File.CreateText("/bitwarden/env/uid.env")) { }
```
Comment (kspearrin) [inline util/Setup/EnvironmentFileBuilder.cs]: This has to stay for windows hosts since the docker file references a uid.env on all hosts

Comment (Mart124) [reply util/Setup/EnvironmentFileBuilder.cs]: Ah yes, I did not think about Win hosts.
I added a test to create the file only if it does not exist.

**Comment by Mart124** [comment]
Thank you 👍
